### PR TITLE
[12.0][MIG] purchase_open_qty

### PR DESCRIPTION
**Pull request by bjeficent**
purchase_open_qty: Migration to version 12.0

### CODE REVIEW

**Review by AaronHForgeFlow — APPROVED** [review]
:+1: Functional review

File: purchase_open_qty/tests/test_purchase_open_qty.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        }
+        self.analytic_account_1 = \
+            analytic_account_model.sudo().create(ac_dict)
+
+        # Purchase Order Num 1
+        po_dict = {
+            'partner_id': self.partner.id,
+        }
+        self.purchase_order_1 = self.purchase_order_model.create(po_dict)
+        uom_id = prod_model.uom_id.search([
```
Comment (nikul-serpentcs) [inline purchase_open_qty/tests/test_purchase_open_qty.py:41]: @bjeficent in search add limit=1, so no need to [0]

File: purchase_open_qty/models/purchase_order.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        po_lines = po_line_obj.search(
+            [('qty_to_receive', operator, value)])
+        order_ids = po_lines.mapped('order_id')
+        res.append(('id', 'in', order_ids.ids))
+        return res
+
+    qty_to_invoice = fields.Float(compute='_compute_qty_to_invoice',
+                                  search='_search_qty_to_invoice',
+                                  string="Qty to Bill",
+                                  default=0.0)
```
Comment (nikul-serpentcs) [inline purchase_open_qty/models/purchase_order.py]: @bjeficent no need to `default`, It's always `0.0`

File: purchase_open_qty/models/purchase_order.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    qty_to_receive = fields.Float(compute='_compute_qty_to_receive',
+                                  search='_search_qty_to_receive',
+                                  string="Qty to Receive",
+                                  default=0.0)
```
Comment (nikul-serpentcs) [inline purchase_open_qty/models/purchase_order.py]: same here

**Review by nikul-serpentcs — CHANGES_REQUESTED** [review]
Minor change

**Comment by bjeficent** [comment]
Updates done. @nikul-serpentcs could you update your review please?

**Review by nikul-serpentcs — APPROVED** [review]
LGTM :+1:

File: purchase_open_qty/models/purchase_order.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+from odoo import api, fields, models
+import odoo.addons.decimal_precision as dp
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = 'purchase.order.line'
+
+    @api.depends('product_qty', 'qty_invoiced',
+                 'invoice_lines.invoice_id.state',
```
Comment (rousseldenis) [inline purchase_open_qty/models/purchase_order.py]: @bjeficent I think the depends should be reviewed.  'invoice_lines.invoice_id.state', 'order_id.state' and 'move_ids.state' are not relevant. 'product_id.purchase_method' should be added

File: purchase_open_qty/models/purchase_order.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            if line.product_id.purchase_method == 'receive':
+                qty = line.qty_received - line.qty_invoiced
+                if qty >= 0.0:
+                    line.qty_to_invoice = qty
+                else:
+                    line.qty_to_invoice = 0.0
+            else:
+                line.qty_to_invoice = line.product_qty - line.qty_invoiced
+
+    @api.depends('order_id.state', 'move_ids.state', 'product_qty')
```
Comment (rousseldenis) [inline purchase_open_qty/models/purchase_order.py]: @bjeficent Same as above. You should review this as some fields should be added (product_uom, ...)

File: purchase_open_qty/models/purchase_order.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    _inherit = 'purchase.order.line'
+
+    @api.depends('product_qty', 'qty_invoiced',
+                 'invoice_lines.invoice_id.state',
+                 'order_id.state', 'move_ids.state', 'qty_received')
+    def _compute_qty_to_invoice(self):
+        for line in self:
+            if line.product_id.purchase_method == 'receive':
+                qty = line.qty_received - line.qty_invoiced
+                if qty >= 0.0:
```
Comment (rousseldenis) [inline purchase_open_qty/models/purchase_order.py]: @bjeficent Use of float_compare is mandatory here

File: purchase_open_qty/models/purchase_order.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    def _compute_qty_to_receive(self):
+        for po in self:
+            po.qty_to_receive = sum(po.mapped('order_line.qty_to_receive'))
+
+    @api.model
+    def _search_qty_to_invoice(self, operator, value):
+        po_line_obj = self.env['purchase.order.line']
+        res = []
+        po_lines = po_line_obj.search(
+            [('qty_to_invoice', operator, value)])
```
Comment (rousseldenis) [inline purchase_open_qty/models/purchase_order.py]: @bjeficent I think this is not correct as the order field is the sum of lines. If there are 5 lines of qty_invoice == 1, the qty_to_invoice in order == 5. But if you search with qty  == 5, you return nothing... Moreover, the test case is not complete on that.

File: purchase_open_qty/models/purchase_order.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        order_ids = po_lines.mapped('order_id')
+        res.append(('id', 'in', order_ids.ids))
+        return res
+
+    @api.model
+    def _search_qty_to_receive(self, operator, value):
+        po_line_obj = self.env['purchase.order.line']
+        res = []
+        po_lines = po_line_obj.search(
```
Comment (rousseldenis) [inline purchase_open_qty/models/purchase_order.py:87]: @bjeficent Same as above

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review. Some changes needed

**Comment by bjeficent** [comment]
@rousseldenis Thanks for your review, i have updated the code but i am not sure if those changes are what you expected. Could you update your review please?

**Comment by bjeficent** [comment]
Ping @rousseldenis. Could you update your review please?

**Comment by AaronHForgeFlow** [comment]
Ping @rousseldenis

**Comment by AaronHForgeFlow** [comment]
@rousseldenis Do you think your comments here have been attended?

File: purchase_open_qty/__manifest__.py [hunk 9]
```diff
@@ -6,12 +6,12 @@
...
     "summary": "Allows to identify the purchase orders that have quantities "
                "pending to invoice or to receive.",
-    "version": "11.0.1.1.0",
+    "version": "12.0.1.1.0",
     "author": "Eficent, "
               "Odoo Community Association (OCA)",
     "website": "https://github.com/OCA/purchase-workflow",
     "category": "Purchases",
-    "depends": ["purchase"],
+    "depends": ["purchase_stock"],
```
Comment (MiquelRForgeFlow) [inline purchase_open_qty/__manifest__.py:14]: I am not sure if this change is needed.

**Review by MiquelRForgeFlow — APPROVED** [review]
LGTM :+1:

File: purchase_open_qty/__manifest__.py (hunk 9 above)
Comment (AaronHForgeFlow) [reply purchase_open_qty/__manifest__.py:14]: It is, it fails to install if purchase_stock is not installed

**Review by AdriaGForgeFlow — APPROVED** [review]
Functional test :+1: 

@rousseldenis Could you update review please?

File: purchase_open_qty/models/purchase_order.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+import odoo.addons.decimal_precision as dp
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = 'purchase.order.line'
+
+    @api.depends('product_qty', 'qty_invoiced', 'product_id.purchase_method',
+                 'qty_received')
+    def _compute_qty_to_invoice(self):
+        for line in self:
```
Comment (rousseldenis) [inline purchase_open_qty/models/purchase_order.py:18]:
```suggestion
        precision = self.env['decimal.precision'].precision_get('Product Unit of Measure')
        for line in self:
```

File: purchase_open_qty/models/purchase_order.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+            if line.product_id.purchase_method == 'receive':
+                qty = line.qty_received - line.qty_invoiced
+                if float_compare(qty, 0.0, precision_digits=2.0):
```
Comment (rousseldenis) [inline purchase_open_qty/models/purchase_order.py]:
```suggestion
                if float_compare(qty, 0.0, precision_digits=precision):
```

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Seems great. Missing one

**Review by AaronHForgeFlow — APPROVED** [review]

**Review by rousseldenis — APPROVED** [review]

**Comment by MiquelRForgeFlow** [comment]
@ageficent Could you add https://github.com/OCA/purchase-workflow/pull/712?

**Comment by AdriaGForgeFlow** [comment]
> @ageficent Could you add #712?

Done :+1:

**Review by JordiBForgeFlow — APPROVED** [review]

**Pull request closed without merging**